examples/example_dotnet/WinSparkleDemoForm.cs

```diff
@@ -0,0 +1,155 @@
+﻿using System;
+using System.Windows.Forms;
+using WinSparkleDotNet;
+
+namespace example_dotnet
+{
+    public partial class WinSparkleDemoForm : Form
+    {
+        private readonly WinSparkleNet _sparkleNet;
+
+        public WinSparkleDemoForm()
+        {
+            InitializeComponent();
+            try
+            {
+                _sparkleNet = new WinSparkleNet();
+            }
+            catch (Exception ex)
+            {
+                MessageBox.Show(@"Error was : "+ex, @"Error during SparkleNet instantiation",
+                    MessageBoxButtons.OK, MessageBoxIcon.Error);
+                return;
+            }
+            //Callback
+            _sparkleNet.SetCanShutdownCallback(this.CanShutDownCallback);
+            _sparkleNet.SetShutdownRequestCallback(this.ShutDownRequestCallback);
+            _sparkleNet.SetDidFindUpdateCallback(this.DidFindAnUpdateCallback);
+            _sparkleNet.SetDidNotFindUpdateCallback(this.DidNotFindAnUpdateCallback);
+            _sparkleNet.SetUpdateCancelledCallback(this.UpdateCancelledCallback);
+            _sparkleNet.SetErrorCallback(this.ErrorCallback);
+            //Init
+            _sparkleNet.Initialize();
+            _sparkleNet.CheckForUpdate();
+        }
+
+        private void Form_FormClosed(object sender, FormClosedEventArgs e)
+        {
+            _sparkleNet.Cleanup();
+        }
+
+        private void Form_Load(object sender, System.EventArgs e)
+        {
+            txtb_lastChecked.Text = _sparkleNet.LastCheckTime.ToString("dd'/'MM'/'yyyy HH:mm:ss");
+            txtb_updateInterval.Text = _sparkleNet.UpdateInterval.Seconds.ToString();
+            ckb_autoCheckUpdate.Checked = _sparkleNet.AutomaticCheckForUpdates;
+            llbl_appCastUrl.Text = _sparkleNet.AppCastUrl;
+        }
+
+        private void btn_checkUpdate_Click(object sender, System.EventArgs e)
+        {
+            _sparkleNet.CheckForUpdate();
+        }
+
+        private void btn_checkUpdateNoUI_Click(object sender, EventArgs e)
+        {
+            _sparkleNet.CheckForUpdate(false);
+        }
+
+        private void btn_checkUpdateAndInstall_Click(object sender, EventArgs e)
+        {
+            _sparkleNet.CheckForUpdate(true, true);
+        }
+
+        private void btn_close_Click(object sender, EventArgs e)
+        {
+            this.Close();
+        }
+
+        private bool CanShutDownCallback()
+        {
+            try
+            {
+                var result = MessageBox.Show(@"Here you can simulate if the application can be closed or not to proceed the update", @"CanShutDownCallback",
```

Comment:
It is not appropriate for the `CanShutDownCallback` callback to show any UI, as you did. It simply returns whether it is safe to terminate the app or not (e.g. because the user has a document open with unsaved edits).

The purpose of an example is *not* to educate the developer how to use the API — that’s what the documentation is for. The purpose is to have a  (almost) *real-life example* of the API is used in practice. I.e. it shouldn’t try to execute everything, it shouldn’t have any such message boxes (why not a comment in the first place?!). It should just ordinarily *use* WinSparkle API as a real app would.